dev/UndockedRegFreeWinRT/urfw.cpp

```diff
@@ -0,0 +1,464 @@
+    if (hr == REGDB_E_CLASSNOTREG)
+    {
+        return TrueRoActivateInstance(activatableClassId, instance);
+    }
+    RETURN_IF_FAILED(hr);
+
+    // Activate in current apartment
+    if (location == ActivationLocation::CurrentApartment)
+    {
+        Microsoft::WRL::ComPtr<IActivationFactory> pFactory;
+        RETURN_IF_FAILED(WinRTGetActivationFactory(activatableClassId, __uuidof(IActivationFactory), (void**)&pFactory));
+        return pFactory->ActivateInstance(instance);
+    }
+
+    // Cross apartment MTA activation
+    struct CrossApartmentMTAActData {
+        HSTRING activatableClassId;
+        IStream *stream;
+    };
+
+    CrossApartmentMTAActData cbdata{ activatableClassId };
+    CO_MTA_USAGE_COOKIE mtaUsageCookie;
+    RETURN_IF_FAILED(CoIncrementMTAUsage(&mtaUsageCookie));
+    RETURN_IF_FAILED(EnsureMTAInitialized());
+    Microsoft::WRL::ComPtr<IContextCallback> defaultContext;
+    ComCallData data;
+    data.pUserDefined = &cbdata;
+    RETURN_IF_FAILED(CoGetDefaultContext(APTTYPE_MTA, IID_PPV_ARGS(&defaultContext)));
+    RETURN_IF_FAILED(defaultContext->ContextCallback(
+        [](_In_ ComCallData* pComCallData) -> HRESULT
+        {
+            CrossApartmentMTAActData* data = reinterpret_cast<CrossApartmentMTAActData*>(pComCallData->pUserDefined);
+            Microsoft::WRL::ComPtr<IInspectable> instance;
+            Microsoft::WRL::ComPtr<IActivationFactory> pFactory;
+            RETURN_IF_FAILED(WinRTGetActivationFactory(data->activatableClassId, __uuidof(IActivationFactory), (void**)&pFactory));
+            RETURN_IF_FAILED(pFactory->ActivateInstance(&instance));
+            RETURN_IF_FAILED(CoMarshalInterThreadInterfaceInStream(IID_IInspectable, instance.Get(), &data->stream));
+            return S_OK;
+        },
+        &data, IID_ICallbackWithNoReentrancyToApplicationSTA, 5, nullptr)); // 5 is meaningless.
+    RETURN_IF_FAILED(CoGetInterfaceAndReleaseStream(cbdata.stream, IID_IInspectable, (LPVOID*)instance));
+    return S_OK;
+}
+
+HRESULT WINAPI RoGetActivationFactoryDetour(HSTRING activatableClassId, REFIID iid, void** factory)
+{
+    ActivationLocation location;
+    HRESULT hr = GetActivationLocation(activatableClassId, location);
+    if (hr == REGDB_E_CLASSNOTREG)
+    {
+        return TrueRoGetActivationFactory(activatableClassId, iid, factory);
+    }
+    RETURN_IF_FAILED(hr);
+
+    // Activate in current apartment
+    if (location == ActivationLocation::CurrentApartment)
+    {
+        RETURN_IF_FAILED(WinRTGetActivationFactory(activatableClassId, iid, factory));
+        return S_OK;
+    }
+    // Cross apartment MTA activation
+    struct CrossApartmentMTAActData {
+        HSTRING activatableClassId;
+        IStream* stream;
+    };
+    CrossApartmentMTAActData cbdata{ activatableClassId };
+    CO_MTA_USAGE_COOKIE mtaUsageCookie;
+    RETURN_IF_FAILED(CoIncrementMTAUsage(&mtaUsageCookie));
+    RETURN_IF_FAILED(EnsureMTAInitialized());
```

Comment:
Is this call necessary?  You just called CoIncrementMTAUsage

Reply:
Unforked UndockedRegFreeWinRT

@BenJKuhn @kythant is there some subtle reason here?

Reply:
I'm not aware of why both would be necessary, will defer to @kythant.

Reply:
If CoInrementMTAUsage does not init the MTA, the world everywhere falls apart.  You can rely on this.

Reply:
This came out of the UndockedRegFreeWinRT project. I'd prefer to follow their lead on the URFW behavior. Will follow up with @kythant  and @BenJKuhn re updating URFW with the fixes I made and additional questions (like this) and sync here with any changes made upstream.